tests/dragonfly/sentinel_test.py

```diff
@@ -0,0 +1,195 @@
+        if assert_ok:
+            assert cp.returncode == 0, f"Command failed: {run_args}"
+        return cp
+
+    def wait_ready(self):
+        wait_for(
+            lambda: self.run_cmd(["ping"], sentinel_cmd=False, assert_ok=False),
+            lambda cp:cp.returncode == 0,
+            timeout_sec=10, timeout_msg="Timeout waiting for sentinel to become ready.")
+
+    def master(self, deployment="") -> dict:
+        if deployment == "":
+            deployment = self.default_deployment
+        cp = self.run_cmd(["master", deployment], capture_output=True)
+        return stdout_as_list_of_dicts(cp)[0]
+
+    def slaves(self, deployment="") -> dict:
+        if deployment == "":
+            deployment = self.default_deployment
+        cp = self.run_cmd(["slaves", deployment], capture_output=True)
+        return stdout_as_list_of_dicts(cp)
+
+    def live_master_port(self, deployment=""):
+        if deployment == "":
+            deployment = self.default_deployment
+        cp = self.run_cmd(["get-master-addr-by-name", deployment], capture_output=True)
+        return int(cp.stdout.splitlines()[1])
+
+    def failover(self, deployment=""):
+        if deployment == "":
+            deployment = self.default_deployment
+        self.run_cmd(["failover", deployment,])
+
+
+@pytest.fixture(scope="function") # Sentinel has state which we don't want carried over form test to test.
+def sentinel() -> Sentinel:
+    s = Sentinel()
+    s.start()
+    s.wait_ready()
+    yield s
+    s.stop()
+
+
+@pytest.mark.asyncio
+async def test_failover(df_local_factory, sentinel):
+    master = df_local_factory.create(port=sentinel.initial_master_port)
+    replica = df_local_factory.create(port=master.port + 1)
+
+    master.start()
+    replica.start()
+
+    master_client = aioredis.Redis(port=master.port)
+    replica_client = aioredis.Redis(port=replica.port)
+
+    await replica_client.execute_command("REPLICAOF localhost " + str(master.port))
+
+    assert sentinel.live_master_port() == master.port
+
+    # Verify sentinel picked up replica.
+    await await_for(
+            lambda: sentinel.master(),
+            lambda m: m["num-slaves"] == "1",
+            timeout_sec=15, timeout_msg="Timeout waiting for sentinel to pick up replica.")
+
+    sentinel.failover()
+
+    # Verify sentinel switched.
+    await await_for(
+        lambda: sentinel.live_master_port(),
+        lambda p: p == replica.port,
+        timeout_sec=3, timeout_msg="Timeout waiting for sentinel to report replica as master."
+    )
+    assert sentinel.slaves()[0]["port"] == str(master.port)
+
+    # Verify we can now write to replica and read replicated value from master.
+    await replica_client.set("key", "value")
+    async def get(): return await master_client.get("key") # There is no async lambda
+    await await_for(
+        get,
+        lambda val: val == b"value",
+        10, "Timeout waiting for key to exist in replica."
+    )
```

Comment:
Sometimes generic utils and async don't mix well in Python. You could've used lambdas if you made them return coroutines which are awaited internally. Or not used the util at all:
```py
async with asyncio.timeout(10):
  while (await master_client.get('key')) != b"value":
    await asyncio.sleep(1.0)
```
But it doesn't matter as its just tests

Reply:
OK, returning a coroutine makes sense to me, I'll merge this version and try that later.